File: backend/Services/Parsers/RaddleParser.cs
using System.Text.RegularExpressions;
using backend.Models;

namespace backend.Services.Parsers
{
    public class RaddleParser : BasicScoreResultParser
    {
        private ILogger<RaddleParser> _logger;

        public RaddleParser(ILogger<RaddleParser> logger) : base(logger)
        {
            _logger = logger;
        }

        public override bool GolfScoring => false;
        public override bool CountWinner => true;
        public override string GameName => "Raddle";
        public override string? HelpText => null;
        protected override Regex Parser => new Regex($".+?\\[(?<{ScoreGroup}>.+?)\\].*?{GameName}", RegexOptions.Singleline);
        protected override string? ExtraContent => null;
        public override string Url => "https://raddle.quest";

        protected override DailyResult SetScore(DailyResult dailyResult, Match parserResults)
        {
            if (!parserResults.Groups.ContainsKey(ScoreGroup))
            {
                return dailyResult;
            }

            var stringScore = parserResults.Groups[ScoreGroup].Value;

            if (stringScore == "💯")
            {
                dailyResult.Score = 100;
            }

            if (stringScore.EndsWith('%'))
            {
                stringScore = stringScore.Replace("%", string.Empty);
            }

            if (int.TryParse(stringScore, out var score))
            {
                dailyResult.Score = score;
            }

            return dailyResult;
        }
    }
}

File: lambdas/src/parsers/raddle-parser.ts
import { DailyResult } from '../models/daily-result';
import { BasicScoreResultParser } from './base/basic-score-parser';
import { getGroup, hasGroup } from './base/result-parser';

export class RaddleParser extends BasicScoreResultParser {
  readonly countWinner = true;
  readonly gameName = 'Raddle';
  readonly golfScoring = false;
  readonly helpText = null;
  protected readonly extraContent = null;
  readonly url = 'https://raddle.quest';
  
  // Matches "TOCO → TOUCAN [💯]\nRaddle #342..." - score is in brackets before "Raddle"
  protected readonly parser = new RegExp(
    `.+?\\[(?<${BasicScoreResultParser.SCORE_GROUP}>.+?)\\].*?${this.gameName}`,
    's'
  );
  
  protected setScore(dailyResult: DailyResult, match: RegExpMatchArray): DailyResult {
    if (!hasGroup(match, BasicScoreResultParser.SCORE_GROUP)) {
      return dailyResult;
    }
    
    let stringScore = getGroup(match, BasicScoreResultParser.SCORE_GROUP) || '';
    
    // Handle 💯 emoji as 100
    if (stringScore === '💯') {
      dailyResult.score = 100;
    }
    
    // Handle percentage suffix
    if (stringScore.endsWith('%')) {
      stringScore = stringScore.replace('%', '');
    }
    
    const score = parseInt(stringScore, 10);
    if (!isNaN(score)) {
      dailyResult.score = score;
    }
    
    return dailyResult;
  }
}
